contracts/staking/PoSValidatorManager.sol

```diff
@@ -130,4 +140,155 @@ abstract contract PoSValidatorManager is IPoSValidatorManager, ValidatorManager
+
+        // Ensure the validation period is active
+        Validator memory validator = _getValidator(validationID);
+        require(
+            validator.status == ValidatorStatus.Active, "ValidatorManager: Validator not active"
+        );
+
+        // Ensure the delegator is not already registered
+        require(
+            $._delegatorStakes[validationID][delegator].weight == 0,
+            "ValidatorManager: Delegator already registered"
+        );
+
+        _checkAndUpdateChurnTracker(weight);
+
+        // Submit the message to the Warp precompile.
+        bytes memory setValidatorWeightPayload = ValidatorMessages
+            .packSetSubnetValidatorWeightMessage(
+            validationID, _getAndIncrementNonce(validationID), validator.weight + weight
+        );
+        $._pendingRegisterDelegatorMessages[validationID][delegator] = setValidatorWeightPayload;
+        bytes32 messageID = WARP_MESSENGER.sendWarpMessage(setValidatorWeightPayload);
+
+        // Store the delegator information
+        $._delegatorStakes[validationID][delegator] = Delegator({
+            weight: weight,
+            startedAt: uint64(block.timestamp),
+            endedAt: 0,
+            status: ValidatorStatus.PendingAdded
+        });
+
+        emit DelegatorAdded(validationID, messageID, delegator, weight, block.timestamp);
+    }
+
+    function resendDelegatorRegistration(bytes32 validationID, address delegator) external {
+        PoSValidatorManagerStorage storage $ = _getPoSValidatorManagerStorage();
+        require(
+            $._pendingRegisterDelegatorMessages[validationID][delegator].length > 0
+                && $._delegatorStakes[validationID][delegator].status == ValidatorStatus.PendingAdded,
+            "PoSValidatorManager: Delegator registration not pending"
+        );
+
+        // Submit the message to the Warp precompile.
+        WARP_MESSENGER.sendWarpMessage($._pendingRegisterDelegatorMessages[validationID][delegator]);
+    }
+
+    function completeDelegatorRegistration(uint32 messageIndex, address delegator) external {
+        PoSValidatorManagerStorage storage $ = _getPoSValidatorManagerStorage();
+
+        // Unpack the Warp message
+        WarpMessage memory warpMessage = _getPChainWarpMessage(messageIndex);
+        (bytes32 validationID, uint64 nonce, uint64 weight) =
+            ValidatorMessages.unpackSubnetValidatorWeightUpdateMessage(warpMessage.payload);
+        require(
+            $._pendingRegisterDelegatorMessages[validationID][delegator].length > 0
+                && $._delegatorStakes[validationID][delegator].status == ValidatorStatus.PendingAdded,
+            "PoSValidatorManager: Delegator registration not pending"
+        );
+        delete $._pendingRegisterDelegatorMessages[validationID][delegator];
+
+        // Update the validator weight and delegator status
+        Validator memory validator = _getValidator(validationID);
+        require(
```

Comment:
I don't think this check is valid if multiple delegations are pending for the same validation ID, and we receive the Warp message for the last delegation

Reply:
Good point. Since we have the nonce, we should be able to remove this check altogether.